libraries/type-length-value/src/state.rs

```diff
@@ -300,37 +357,87 @@ impl<'data> TlvStateMut<'data> {
         Ok(&mut self.data[value_start..value_end])
     }
 
-    /// Packs the default TLV data into the first open slot in the data buffer.
-    /// If extension is already found in the buffer, it returns an error.
-    pub fn init_value<V: SplDiscriminate + Pod + Default>(
+    /// Unpack a portion of the TLV data as mutable bytes for the first entry
+    /// found
+    pub fn get_first_bytes_mut<V: SplDiscriminate>(&mut self) -> Result<&mut [u8], ProgramError> {
+        self.get_bytes_mut::<V>(0)
+    }
+
+    fn init_value_at_entry<V: SplDiscriminate + Pod + Default>(
         &mut self,
-    ) -> Result<&mut V, ProgramError> {
+        entry_number: Option<usize>,
+    ) -> Result<(&mut V, usize), ProgramError> {
         let length = size_of::<V>();
-        let buffer = self.alloc::<V>(length)?;
+        let (buffer, entry_number) = self.alloc::<V>(length, entry_number)?;
         let extension_ref = pod_from_bytes_mut::<V>(buffer)?;
         *extension_ref = V::default();
+        Ok((extension_ref, entry_number))
+    }
+
+    /// Packs the default TLV data into the first open slot in the data buffer.
+    /// Does not check for duplicates. Will add a new entry to the next open
+    /// slot provided there is enough space.
+    pub fn init_value<V: SplDiscriminate + Pod + Default>(
+        &mut self,
+    ) -> Result<(&mut V, usize), ProgramError> {
+        self.init_value_at_entry::<V>(None)
+    }
+
+    /// Packs the default TLV data into the first open slot in the data buffer.
+    /// If extension is already found in the buffer, it returns an error.
+    pub fn init_value_unique<V: SplDiscriminate + Pod + Default>(
+        &mut self,
+    ) -> Result<&mut V, ProgramError> {
+        let (extension_ref, _) = self.init_value_at_entry::<V>(Some(0))?;
         Ok(extension_ref)
     }
 
-    /// Packs a variable-length value into its appropriate data segment. Assumes that
-    /// space has already been allocated for the given type
-    pub fn pack_variable_len_value<V: SplDiscriminate + VariableLenPack>(
+    /// Packs a variable-length value into its appropriate data segment. Assumes
+    /// that space has already been allocated for the given type
+    fn pack_variable_len_value_at_entry<V: SplDiscriminate + VariableLenPack>(
         &mut self,
         value: &V,
+        entry_number: usize,
     ) -> Result<(), ProgramError> {
-        let data = self.get_bytes_mut::<V>()?;
+        let data = self.get_bytes_mut::<V>(entry_number)?;
         // NOTE: Do *not* use `pack`, since the length check will cause
         // reallocations to smaller sizes to fail
         value.pack_into_slice(data)
     }
 
+    /// Packs a variable-length value into its appropriate data segment, where
+    /// repeating discriminators _are_ allowed
+    pub fn pack_variable_len_value<V: SplDiscriminate + VariableLenPack>(
```

Comment:
naming nit: this one is less important since it has the new parameter, but better to be clear
```suggestion
    pub fn pack_variable_len_value_with_repetition<V: SplDiscriminate + VariableLenPack>(
```